aten/src/ATen/native/cpu/SoftMaxKernel.cpp

```diff
@@ -23,80 +23,88 @@
+  // machine's L1D cache size.
+  int64_t CHUNK_SIZE = std::max(
+      (int64_t)1,
+      (int64_t)(at::internal::GRAIN_SIZE / (sizeof(scalar_t) * dim_size)));
 
-  parallel_for(
-      0,
-      outer_size,
-      grain_size,
-      [&](int64_t begin, int64_t end) {
-        // NOLINTNEXTLINE(modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays)
-        scalar_t tmp_sum_scalar[CHUNK_SIZE];
-        // NOLINTNEXTLINE(modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays)
-        scalar_t max_input_arr[CHUNK_SIZE];
-        for (int64_t ii = begin; ii < end; ii += CHUNK_SIZE) {
-          int64_t loop_end = CHUNK_SIZE;
-          if (ii + CHUNK_SIZE > end)
-            loop_end = end - ii;
-          for (const auto j : c10::irange(loop_end)) {
-            int64_t i = ii + j;
-            scalar_t* input_data = input_data_base + i * dim_size;
-            max_input_arr[j] = vec::reduce_all<scalar_t>(
-                [](Vec& x, Vec& y) { return vec::maximum(x, y); },
-                input_data,
-                dim_size);
-          }
-          for (const auto j : c10::irange(loop_end)) {
-            int64_t i = ii + j;
-            scalar_t* input_data = input_data_base + i * dim_size;
-            scalar_t max_input = max_input_arr[j];
-            tmp_sum_scalar[j] = vec::map_reduce_all<scalar_t>(
-                [max_input](Vec x) { return (x - Vec(max_input)).exp(); },
-                [](Vec x, Vec y) { return x + y; },
-                input_data,
-                dim_size);
-          }
-          // See [Note AVX-SSE transitions] for why this should call the
-          // vectorized version (aside from perf improvements).
-          vec::map(
-              [](Vec x) { return x.log(); },
-              tmp_sum_scalar,
-              tmp_sum_scalar,
-              loop_end);
-          for (const auto j : c10::irange(loop_end)) {
-            int64_t i = ii + j;
-            scalar_t* input_data = input_data_base + i * dim_size;
-            scalar_t* output_data = output_data_base + i * dim_size;
-            scalar_t tmp_sum = tmp_sum_scalar[j];
-            scalar_t max_input = max_input_arr[j];
-
-            // It's necessary to keep the order of the operations below.
-            // In some cases that input is large digits and the difference
-            // is small, if we compute `max_input` plus `tmp_sum` before,
-            // there would be a numerical problem. See an example in
-            // https://github.com/pytorch/pytorch/issues/11752#issuecomment-422883379
-            vec::map(
-                [tmp_sum, max_input](Vec x) { return x - Vec(max_input) - Vec(tmp_sum); },
-                output_data,
-                input_data,
-                dim_size);
-          }
-        }
-      });
+  // we assume that at::GRAIN_SIZE is an appropriate grain-size.
+  int64_t grain_size = std::max(
+      (Vec::size() * internal::GRAIN_SIZE) / (16 * dim_size), 
```

Comment:
This is roughly `max(4* Vec::size() * CHUNK_SIZE, 1)`. Do we want to make `grain_size` depend on the dtype size as well or nah? and if so, should it depend linearly or inversely?

Reply:
`Vec::size()` has an inverse relationship with dtype size, so here, `grain_size` inversely depends upon the dtype size.

I guess `Vec::size()` is multiplied with `internal::GRAIN_SIZE` to reflect the fact that the effective grain-size is larger due to vectorization (computation on more elements can be done with the same number of instructions). 
However, I guess when [`internal::GRAIN_SIZE`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIterator.h#L80-L86) had been added to the codebase, SSE or AVX vectorization would already have been present, so should `GRAIN_SIZE` be construed as _`minimum number of elements that warrant parallelization (with multi-threading) when vectorization type is SSE/AVX2`_? If so, then we should probably divide `Vec::size()` here with some appropriate number as well.

Thanks!

Reply:
@peterbell10 may have more context to answer this one

Reply:
> Note that a few lines below `_vec_softmax_lastdim` also uses the same heuristic, so when we settle for a proper one, we should probably also update that one.

Thanks again for your inputs, @lezcano! I just copy-pasted that one in `_vec_log_softmax_lastdim`.

Reply:
> I guess Vec::size() is multiplied with internal::GRAIN_SIZE to reflect the fact that the effective grain-size is larger due to vectorization 

The code on master actually divides by `Vec::size()` (implicitly through `CHUNK_SIZE`) so you're the one introducing the multiply by `Vec::size()`.  I can't see any reason to do it the previous way though. tbh I think `GRAIN_SIZE / dim_size` is enough good enough without detailed benchmarking.

The same default grain size is used on non-vectorized code as well, so I don't think it relates to any specific vector instruction set. 

TBH I think it's a mistake that the existing code divides `GRAIN_SIZE` by `Vec::size()`.

> should GRAIN_SIZE be construed as minimum number of elements that warrant parallelization (with multi-threading) when vectorization type is SSE/AVX2? 

The same grain size is currently used throughout pytorch for non-vectorized loops as well, so I wouldn't think about it too deeply.